kratos/tensor_adaptors/tensor_adaptor.h

```diff
@@ -0,0 +1,142 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+#pragma once
+
+// System includes
+#include <string>
+#include <atomic>
+
+// External includes
+#include <span/span.hpp>
+
+// Project includes
+#include "includes/define.h"
+#include "includes/ublas_interface.h"
+#include "intrusive_ptr/intrusive_ptr.hpp"
+
+namespace Kratos {
+
+///@name Kratos Classes
+///@{
+
+/**
+ * @brief Base class or all the tensor adaptor types.
+ */
+template<class TContainerType, class TPrimitiveDataType>
+class KRATOS_API(KRATOS_CORE) TensorAdaptor {
+public:
+
+    ///@name Type definitions
+    ///@{
+
+    using Pointer = Kratos::intrusive_ptr<TensorAdaptor>;
+
+    using ConstPointer = Kratos::intrusive_ptr<const TensorAdaptor>;
+
+    using PrimitiveDataType = TPrimitiveDataType;
+
+    using ContainerType = TContainerType;
+
+    ///@}
+    ///@name Life cycle
+    ///@{
+
+    TensorAdaptor(typename TContainerType::Pointer pContainer) : mpContainer(pContainer) {};
+
+    virtual ~TensorAdaptor() = default;
+
+    ///@}
+    ///@name Public operations
+    ///@{
+
+    /**
+     * @brief Fill the internal data from Kratos data structures
+     */
+    virtual void CollectData() = 0;
+
+    /**
+     * @brief Store internal data to the given TContainerType container.
+     *
+     */
+    virtual void StoreData() = 0;
+
+    typename TContainerType::Pointer GetContainer() const { return mpContainer; }
+
+    DenseVector<TPrimitiveDataType> MoveData() { return std::move(mData); }
+
+    Kratos::span<const TPrimitiveDataType> ViewData() const { return Kratos::span<const TPrimitiveDataType>(mData.data().begin(), mData.data().end()); }
+
+    Kratos::span<TPrimitiveDataType> ViewData() { return Kratos::span<TPrimitiveDataType>(mData.data().begin(), mData.data().end()); }
+
+    /**
+     * @brief Get the Shape of the tensor
+     */
+    std::vector<int> Shape() const { return mShape; };
```

Comment:
i would return a DenseVector<IndexType>. It has much lower overhead for small vectors than std::vector

Reply:
imagine you have an std::vector of size 1

it actually occupies (capacity*TDataType + size + capacity)*8 bytes

a Dense VEctor would occupy (size*Capacity + 1)*8 bytes. It is a big overhead for small vectors

and there is no problem in converting lists to DenseVectors or viceversa

Reply:
what i am telling is that an std::vector allocated "capacity" items of TDataType (capacity, not size!). Also it keeps track of 2  sizes, the "size" and the "capacity" itself. This is the way it allows "fast" push back (the capacity grows faster than the resize)

https://stackoverflow.com/questions/52330010/what-does-stdvector-look-like-in-memory

i would have no issues using something that is not "DeseVector" (for example "small_vector" you cite) if this thing existed in the stl ... but since to the best of my knowledge it does not, i would go for the DenseVector which we already have in Kratos